Automattic/bash-cache-buildkite-plugin
# Add SPM Caching Action

### Summary

This PR adds a caching action for Swift Package Manager to the CI toolkit. Some notes:
- It works for both workspace-based repos like WCiOS, DOiOS/Mac, etc. along with non-Xcode repos like `hostmgr` or Swift packages if they have SPM dependencies. The action automatically detects what kind of repo it is based on where the `Package.resolved` file is located. 
- I used the existing `install_swiftpm_dependencies` action because I didn't see it being used elsewhere. If that's incorrect (@jkmassel & @AliSoftware looks like you both worked on that action in the past), please let me know and I can move SPM caching into a separate action. 
- This is my first time diving into bash scripting. If there are conventions I'm not following or better ways of doing something, please let me know! 

### How to Test
You can see this CI build of Day One iOS/Mac that's using the latest caching code: 

Example Builds: 
- DOiOS/Mac: https://buildkite.com/automattic/day-one-apple/builds/7705
- WCiOS: https://buildkite.com/automattic/woocommerce-ios/builds/11347
- hostmgr: https://buildkite.com/automattic/hostmgr/builds/295

There is now a "Restoring SPM Cache" step: 

<img width="731" alt="Screenshot 2023-03-01 at 3 39 13 PM" src="https://user-images.githubusercontent.com/17955542/222281800-000a000a-122c-4785-bea5-0a6b4c2b1bd2.png">

To try regenerating the cache, you can delete any of the `day-one-spm-cache...` files from the `a8c-ci-cache` and re-run the build to verify that the cache rebuilds.
